Add Circuit Breakers panel concept (like in Lagom)

ihostage opened the pull request:
Concept (and implementation of course 😄 ) inspired by Lagom Framework.
I'm sure it can be very useful not only for Lagom users but also for Play Framework and Akka Platform users.
Telemetry inspired by Akka Projection.

Refs https://github.com/akka/akka/issues/16617

ihostage committed:
Add Circuit Breakers panel concept (like in Lagom)

ignasi35 commented on akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakersPanelInternal.scala:
```diff
@@ -0,0 +1,104 @@
+    val default: Config = config.getConfig("default")
+  }
+
+  private val breakers = new ConcurrentHashMap[String, Option[CircuitBreakerHolder]]
```
I think the Lagom impl is leaky https://github.com/lagom/lagom/issues/3304

patriknw reviewed (commented):
Telemetry for CircuitBreakers is kind of essential so I like the suggestion, but we might need some more iterations on the API...

ihostage committed:
* Panel as extension and rename to Registry
* Move Telemetry from Registry to CircuitBreaker

ihostage commented:
Hi, @patriknw! Sorry for long silence. I had found a time and reworked this draft by your comments. Can you do a yet review?
Summary of changes:
* `CircuitBreakersPanel` implemented as extension and renamed to `CircuitBreakersRegistry`
* `CircuitBreakerTelemetry` is moved from registry to `CircuitBreaker` directly.

patriknw commented on akka-actor/src/main/scala/akka/pattern/CircuitBreakersRegistry.scala:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * Returns the canonical ExtensionId for this Extension
+   */
+  override def lookup: ExtensionId[_ <: Extension] = CircuitBreakersRegistry
```
two more needed for it to work well from Java:
```
  override def get(system: ActorSystem): CircuitBreakersRegistry = super.get(system)

  override def get(system: ClassicActorSystemProvider): CircuitBreakersRegistry = super.get(system)
```

ihostage replied:
Done ✅

patriknw commented on akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala:
```diff
@@ -0,0 +1,157 @@
+ * INTERNAL API
+ */
+@InternalApi private[akka] object CircuitBreakerNoopTelemetry extends CircuitBreakerTelemetry {
+  override def onOpen(): Unit = {}
```
for empty methods returning Unit we typically use `= ()`

ihostage replied:
Done ✅

patriknw reviewed (commented):
this is looking great

ihostage committed:
* Add configuration for all circuit breaker params
* Add tests for identified circuit breaker

ihostage commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -1307,4 +1307,33 @@ akka {
   }
 
+  #//#circuit-breaker-default
+  # Circuit breakers. A child configuration section with the same
```
Thank you, Johan! 👍

ihostage commented:
Hi, @patriknw! 👋 
Thank you for your comments! 👍 
I added one commit with:
* Add configuration for all circuit breaker params
* Add tests for identified circuit breaker

patriknw commented on akka-testkit/src/test/resources/reference.conf:
```diff
@@ -31,4 +31,15 @@ akka {
+        "akka.pattern.CircuitBreakerSpec$AllowException"
+      ]
+    }
+  }
```
I would prefer to include this in `CircuitBreakerSpec` instead. It's too hidden here.

```
class CircuitBreakerSpec extends AkkaSpec("""
  akka.circuit-breaker {
     identified {
       max-failures = 1
       call-timeout = 100 ms
       reset-timeout = 200 ms
       exception-allowlist = [
         "akka.pattern.CircuitBreakerSpec$AllowException"
       ]
     }
   }
   """) {
```

ihostage replied:
Done ✅

patriknw reviewed (commented):
looking good, just a suggestion for the test

ihostage committed:
Use configuration for tests explicitly in test code

patriknw reviewed (approved):
LGTM

ihostage commented:
Thank you, Patrik @patriknw! 🙏 
@johanandren Johan, WDYT?

johanandren reviewed (approved):
LGTM, nice work @ihostage !

nicoga97 commented:
Hi @ihostage, sorry to bother you with this old PR, but I don't find documentation about how to use this circuit breaker metrics they are exposed by Kamon? or how it work, thanks

ihostage commented:
Hi @nicoga97! I'm not sure that Kamon already supports that. But supporting that is very trivial. 
1. Implement trait [CircuitBreakerTelemetry](https://github.com/akka/akka/blob/3dbcf1c8ea0af031476ca0b50f52bea1d1e5f88e/akka-actor/src/main/scala/akka/pattern/internal/CircuitBreakerTelemetry.scala#L13-L24) like as if I did for Lagom (see [CircuitBreakerMetricsProviderImpl](https://github.com/kamon-io/Kamon/blob/a8bef44040c20047de700bc09f608bbdfdea033b/instrumentation/kamon-lagom/src/main/scala/kamon/instrumentation/lagom/CircuitBreakerMetricsProviderImpl.scala#L32-L78) in Kamon Lagom module)
2. Add implementation by next config 
```HOCON
akka.circuit-breaker.telemetry.implementations += com.example.MyMetrics
```

nicoga97 commented:
@ihostage Thanks for the fast response! Great seems easy!
